opencga-catalog/src/main/java/org/opencb/opencga/catalog/managers/UserManager.java

```diff
@@ -737,12 +745,37 @@ public AuthenticationResponse login(String organizationId, String username, Stri
             }
         }
 
-        OpenCGAResult<User> userOpenCGAResult = getUserDBAdaptor(organizationId).get(username, INCLUDE_ACCOUNT);
+        OpenCGAResult<User> userOpenCGAResult = getUserDBAdaptor(organizationId).get(username, INCLUDE_ACCOUNT_AND_INTERNAL);
         if (userOpenCGAResult.getNumResults() == 1) {
+            User user = userOpenCGAResult.first();
+            if (!ParamConstants.ADMIN_ORGANIZATION.equals(organizationId)) {
+                // Check user is not banned or account is expired
+                if (UserStatus.BANNED.equals(user.getInternal().getStatus().getId())) {
+                    throw CatalogAuthenticationException.userIsBanned(username);
+                }
+                Date date = TimeUtils.toDate(user.getAccount().getExpirationDate());
+                if (date == null) {
+                    throw new CatalogException("Unexpected null expiration date for user '" + username + "'.");
+                }
+                if (date.before(new Date())) {
+                    throw CatalogAuthenticationException.accountIsExpired(username, user.getAccount().getExpirationDate());
+                }
+            }
             authId = userOpenCGAResult.first().getAccount().getAuthentication().getId();
             try {
                 response = authenticationFactory.authenticate(organizationId, authId, username, password);
             } catch (CatalogAuthenticationException e) {
+                if (!ParamConstants.ADMIN_ORGANIZATION.equals(organizationId)) {
+                    // We can only lock the account if it is not the root user
+                    int failedAttempts = userOpenCGAResult.first().getInternal().getFailedAttempts();
+                    ObjectMap updateParams = new ObjectMap(UserDBAdaptor.QueryParams.INTERNAL_FAILED_ATTEMPTS.key(), failedAttempts + 1);
+                    if (failedAttempts >= 4) {
+                        // Ban the account
+                        updateParams.append(UserDBAdaptor.QueryParams.INTERNAL_STATUS_ID.key(), UserStatus.BANNED);
+                    }
+                    getUserDBAdaptor(organizationId).update(username, updateParams);
```

Comment:
Do we want to keep updating the user's document even if the user is already banned or its account expired?

Reply:
Yes, that will let us know how many times the user has attempted to log in and when was the last time. I know it is in the audit but it is also good having that info in the User account.